src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,487 @@
+}
+
+CQubitManager::~CQubitManager()
+{
+    if (sharedQubitStatusArray != nullptr)
+    {
+        delete[] sharedQubitStatusArray;
+        sharedQubitStatusArray = nullptr;
+    }
+    // freeQubitsInAreas - direct member of the class, no need to delete.
+}
+
+// Although it is not necessary to pass area IDs to these functions, such support may be added for extra checks.
+void CQubitManager::StartRestrictedReuseArea()
+{
+    RestrictedReuseArea newArea;
+    freeQubitsInAreas.PushToBack(newArea);
+}
+
+void CQubitManager::NextRestrictedReuseSegment()
+{
+    FailIf(freeQubitsInAreas.Count() <= 0, "Internal error! No reuse areas.");
+    FailIf(freeQubitsInAreas.Count() == 1, "NextRestrictedReuseSegment() without an active area.");
+    RestrictedReuseArea& currentArea = freeQubitsInAreas.PeekBack();
+    // When new segment starts, reuse of all free qubits in the current area becomes prohibited.
+    currentArea.FreeQubitsReuseProhibited.MoveAllQubitsFrom(currentArea.FreeQubitsReuseAllowed, sharedQubitStatusArray);
+}
+
+void CQubitManager::EndRestrictedReuseArea()
+{
+    FailIf(freeQubitsInAreas.Count() < 2, "EndRestrictedReuseArea() without an active area.");
+    RestrictedReuseArea areaAboutToEnd = freeQubitsInAreas.PopFromBack();
+    RestrictedReuseArea& containingArea = freeQubitsInAreas.PeekBack();
+    // When area ends, reuse of all free qubits from this area becomes allowed.
+    containingArea.FreeQubitsReuseAllowed.MoveAllQubitsFrom(areaAboutToEnd.FreeQubitsReuseProhibited, sharedQubitStatusArray);
+    containingArea.FreeQubitsReuseAllowed.MoveAllQubitsFrom(areaAboutToEnd.FreeQubitsReuseAllowed, sharedQubitStatusArray);
+}
+
+Qubit CQubitManager::Allocate()
+{
+    QubitIdType newQubitId = AllocateQubitId();
+    if (newQubitId == NoneMarker && mayExtendCapacity)
+    {
+        QubitIdType newQubitCapacity = qubitCapacity * 2;
+        FailIf(newQubitCapacity <= qubitCapacity, "Cannot extend capacity.");
+        EnsureCapacity(newQubitCapacity);
+        newQubitId = AllocateQubitId();
+    }
+    FailIf(newQubitId == NoneMarker, "Not enough qubits.");
+    ChangeStatusToAllocated(newQubitId);
+    return CreateQubitObject(newQubitId);
+}
+
+void CQubitManager::Allocate(Qubit* qubitsToAllocate, int qubitCountToAllocate)
+{
+    FailIf(qubitCountToAllocate < 0, "Attempt to allocate negative number of qubits.");
+    if (qubitCountToAllocate == 0)
+    {
+        return;
+    }
+    FailIf(qubitsToAllocate == nullptr, "No array provided for qubits to be allocated.");
+
+    // Consider optimization for initial allocation of a large array at once
+    for (int i = 0; i < qubitCountToAllocate; i++)
+    {
+        QubitIdType newQubitId = AllocateQubitId();
+        if (newQubitId == NoneMarker)
+        {
+            for (int k = 0; k < i; k++)
+            {
+                Release(qubitsToAllocate[k]);
+            }
+            FailNow("Not enough qubits.");
+            return;
+        }
```

Comment:
Just curious, can this be optimized out by comparing `qubitCountToAllocate > freeQubitCount`? (before the `for` loop)

Reply:
Not really. First, we would have to have this check for robustness anyway. So we cannot just remove the check. Also even if qubitCountToAllocate > freeQubitCount, we may still be able to allocate if we are allowed to extend capacity.